src/plugins/data/public/antlr/shared/utils.ts

```diff
@@ -4,81 +4,85 @@
+  dataSourceRequestHandler: ({
     dataSourceId,
     title,
   }: IDataSourceRequestHandlerParams) => Promise<any>,
-  defaultReuqstHandler: () => any
+  defaultRequestHandler: () => Promise<any>
 ) =>
-  connectionsService.getSelectedConnection$().pipe(
+  dataSetManager.getUpdates$().pipe(
+    startWith(dataSetManager.getDataSet()),
     distinctUntilChanged(),
-    switchMap((connection: any) => {
-      if (connection === undefined) {
-        return from(defaultReuqstHandler());
+    switchMap((dataSet) => {
+      if (!dataSet) {
+        return from(defaultRequestHandler());
       }
-      const dataSourceId = connection?.dataSource?.id;
-      const title = connection?.attributes?.title;
-      return from(dataSourceReuqstHandler({ dataSourceId, title }));
+      const dataSourceId = dataSet?.dataSourceRef?.id;
+      const title = dataSet?.dataSourceRef?.name;
+      return from(dataSourceRequestHandler({ dataSourceId, title }));
     })
   );
 
+const fetchFromAPI = async (api: any, body: string) => {
+  try {
+    return await api.http.fetch({
+      method: 'POST',
+      path: '/api/enhancements/search/sql',
+      body,
+    });
+  } catch (err) {
+    // TODO: pipe error to UI
+    return Promise.reject(err);
+  }
+};
+
+// Generic fetchData function
 export const fetchData = (
   tables: string[],
   queryFormatter: (table: string, dataSourceId?: string, title?: string) => any,
   api: any,
-  connectionService: any
-): Promise<any[]> => {
+  dataSetManager: DataSetManager
+) => {
   return new Promise((resolve, reject) => {
     getRawSuggestionData$(
-      connectionService,
+      dataSetManager,
       ({ dataSourceId, title }) => {
-        return Promise.all(
-          tables.map(async (table) => {
-            const body = JSON.stringify(queryFormatter(table, dataSourceId, title));
-            return api.http.fetch({
-              method: 'POST',
-              path: '/api/enhancements/search/sql',
-              body,
-            });
-          })
-        );
+        const requests = tables.map(async (table) => {
+          const body = JSON.stringify(queryFormatter(table, dataSourceId, title));
+          return fetchFromAPI(api, body);
+        });
+        return Promise.all(requests);
       },
       () => {
-        return Promise.all(
-          tables.map(async (table) => {
-            const body = JSON.stringify(queryFormatter(table));
-            return api.http.fetch({
-              method: 'POST',
-              path: '/api/enhancements/search/sql',
-              body,
-            });
-          })
-        );
+        const requests = tables.map(async (table) => {
+          const body = JSON.stringify(queryFormatter(table));
+          return fetchFromAPI(api, body);
+        });
+        return Promise.all(requests);
       }
     ).subscribe({
       next: (dataFrames: any) => resolve(dataFrames),
-      error: (err: any) => {
+      error: (err: Error) => {
```

Comment:
seems like this resolves and rejects immediately and only once, do you need to make getRawSuggestionData an observable? and it would be better to avoid `any` for easier maintenance